include/seqan3/core/metafunction/range.hpp

```diff
@@ -0,0 +1,282 @@
+ * \tparam t The type to recurse on; must have `ranges::value_type_t<rng_t>`
+ *
+ * \details
+ *
+ * Attention, this metafunction implicitly removes cv-qualifiers on the all value_types except the one returned.
+ */
+template <typename t>
+//!\cond
+    requires requires (t) { typename value_type_t<remove_cvref_t<t>>; }
+//!\endcond
+struct innermost_value_type
+{
+    //!\brief The return type (recursion not shown).
+    using type = value_type_t<remove_cvref_t<t>>;
+};
+
+//!\cond
+template <typename t>
+    requires requires (t)
+    {
+        typename value_type_t<remove_cvref_t<t>>;
+        typename value_type_t<remove_cvref_t<value_type_t<remove_cvref_t<t>>>>;
+    }
+struct innermost_value_type<t>
+{
+    using type = typename innermost_value_type<value_type_t<remove_cvref_t<t>>>::type;
+};
+//!\endcond
+
+//!\brief Shortcut for seqan3::innermost_value_type.
+template <typename t>
+using innermost_value_type_t = typename innermost_value_type<t>::type;
+
+// ----------------------------------------------------------------------------
+// dimension_v
+// ----------------------------------------------------------------------------
+
+/*!\brief Returns the number of times you can call `seqan3::value_type_t` recursively on t [Value metafunction].
+ * \tparam t The type to be queried; must resolve `seqan3::value_type_t` at least once.
+ *
+ * \details
+ *
+ * Attention, this metafunction implicitly removes cv-qualifiers and reference from the types it recurses on and
+ * returns.
+ */
+template <typename t>
+//!\cond
+    requires requires (t) { typename value_type_t<remove_cvref_t<t>>; }
+//!\endcond
+constexpr size_t dimension_v = 1;
+
+//!\cond
+template <typename t>
+    requires requires (t)
+    {
+        typename value_type_t<remove_cvref_t<t>>;
+        typename value_type_t<value_type_t<remove_cvref_t<t>>>;
+    }
+constexpr size_t dimension_v<t> = dimension_v<value_type_t<remove_cvref_t<t>>> + 1;
+//!\endcond
+
+// ----------------------------------------------------------------------------
+// compatible_concept
+// ----------------------------------------------------------------------------
+
+/*!\interface seqan3::compatible_concept <>
+ * \brief Two types are "compatibe" if their seqan3::dimension_v and their seqan3::innermost_value_type_t are
+ * the same.
+ *
+ * \details
+ *
+ * ```cpp
+ * // these evaluate to true:
+ * static_assert(seqan3::compatible_concept<std::string,              std::vector<char>>);
+ * static_assert(seqan3::compatible_concept<std::vector<std::string>, std::vector<std::vector<char>>>);
+ * ```
+ *
+ * Attention, this metafunction implicitly removes cv-qualifiers and reference from the types it recurses on and
+ * compares.
+ */
+//!\cond
+template <typename t1, typename t2>
+concept bool compatible_concept = requires (t1, t2)
+{
+    requires (dimension_v<t1> == dimension_v<t2>);
+
+    requires std::is_same_v<innermost_value_type_t<t1>, innermost_value_type_t<t2>>; //TODO change to common_reference?
```

Comment:
open todo. Can you explain what the difference would be?

Reply:
I am not yet convinced of this fully. I will delete the TODO.

>  Can you explain what the difference would be?

One compares the actual value_types, the other allows convertible types, too.